---
title: "MMRFBiolinks Clinical data"
package: MMRFBiolinks
output:
  BiocStyle::html_document
vignette: |
  %\VignetteIndexEntry{4. Clinical data}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r setup}
library(MMRFBiolinks)



```


```{r, echo = FALSE,hide=TRUE, message=FALSE,warning=FALSE}
devtools::load_all(".")
```



```{r message=FALSE, warning=FALSE, include=FALSE}
library(TCGAbiolinks)
library(SummarizedExperiment)
library(dplyr)
library(DT)
```


<div class="panel panel-info">
<div class="panel-heading">Different sources of clinical data</div>
<div class="panel-body">

MMRF-Compass clinical data can be retrieved from two different sources:

    
 - **GDC Clinical information**: MMRF-Commpass clinical information download from GDC database;
 - **MMRF-Commpass Clinical information**: Clinical information downloaded directly from MMRF-Commpass Researcher Gateway has more information compared to the previous dataset (e.g Best overall response). So the previous one is only a subset of this last. The user can download this dataset from MMRF-Commpass Researcher Gateway and import it as a dataframe into R environment.



</div>
</div>

# Get MMRF-COMMPASS clinical data at GDC Data Portal
## MMRFGDC\_QueryClinic
In this example we will fetch clinical GDC data.

```{r results='hide', echo=TRUE, message=FALSE, warning=FALSE}
clinical <- MMRFGDC_QueryClinic(type = "clinical")
```

```{r  echo=TRUE, message=FALSE, warning=FALSE}
clinical %>% head %>% datatable(options = list(scrollX = TRUE, keys = TRUE))
 

```


# Get MMRF-COMMPASS data at MMRF-COMMPASS Research Gateway

In this example we will fetch clinical MMRF-COMMPASS Clinical information data form MMRF-COMMPASS Research Gateway. Once the dataset (.csv file) is imported into our own R environment, it is can be explored  as a R-dataframe.


```{r results = 'hide', message=FALSE, warning=FALSE, eval = F}

clinMMGateway <- read.csv("~/MMRF_CoMMpass_IA14_STAND_ALONE_TRTRESP.csv")

```

```{r  echo=TRUE, message=FALSE, warning=FALSE}


clinMMGateway %>% datatable(options = list(scrollX = TRUE, keys = TRUE))

```

## MMRFRG\_GetBorInfo
Get Best Overall Response (BOR) type filtered by sample ID (Case 1) or 
Get sample ID filtered by BOR type (Case 2)


```{r results = 'hide', message=FALSE, warning=FALSE, eval = F}

# case 1
 
 IDsamples<-MMRFRG_GetBorInfo(clin.rg=clinMMGateway,bor="PR") 
 
 # case 2
 
   list.samp<- c("MMRF_0001","MMRF_0002",
                 "MMRF_0003","MMRF_0004",
                 "MMRF_0005","MMRF_0006",
                 "MMRF_0007","MMRF_0008",
                 "MMRF_0009","MMRF_0010")
                 
 
 bestOverall<-MMRFRG_GetBorInfo(clin.rg=clinMMGateway,listSamples=list.samp)  



```


## MMRFGDC\_QuerySamples

Retrieve samples identifiers filtered by the selected type sample (case 2) or 
Retrieve samples identifiers filtered by the selected therapy (case 3) or 
Retrieve samples identifiers from both case 1 and case 2.

```{r results = 'hide', message=FALSE, warning=FALSE, eval = F}
therapy<-"Dexamethasone" 

tsample<-"TRBM"
clin<-MMRFGDC_QueryClinic(type = "clinical")

query.mm <- GDCquery(project = "MMRF-COMMPASS", 
                             data.category = "Transcriptome Profiling",
                             data.type = "Gene Expression Quantification",
                             experimental.strategy = "RNA-Seq",
                             workflow.type="HTSeq - FPKM")



IDs1<-MMRFGDC_QuerySamples(query=query.mm,typesample=tsample, clin.mm=clin,therapyname=therapy) #case 1
IDs2<-MMRFGDC_QuerySamples(query=query.mm,typesample=tsample) #case 2 
IDs3<-MMRFGDC_QuerySamples(clin.mm=clin,therapyname=therapy) #case 3


```








